designs/default-zero-value.md

```diff
@@ -0,0 +1,857 @@
+   semantics through the `@default` trait or `@required` trait.
+5. Deprecate the `PrimitiveBoolean`, `PrimitiveShort`, `PrimitiveInteger`,
+   `PrimitiveLong`, `PrimitiveFloat`, and `PrimitiveDouble` shapes from the
+   Smithy prelude. Instead, target their corresponding prelude shapes like
+   `Boolean`, `Short`, `Integer`, etc.
+6. Create new guidance for code generators that allows them to generate
+   non-nullable accessors for structure members marked as `@required` or
+   `@default`.
+
+
+### In Smithy IDL 2.0
+
+1. Remove the `@box` trait from the Smithy prelude. IDL 2.0 models will fail if
+   they use the `@box` trait.
+2. Remove the `PrimitiveBoolean`, `PrimitiveShort`, `PrimitiveInteger`,
+   `PrimitiveLong`, `PrimitiveFloat`, and `PrimitiveDouble` shapes from the
+   Smithy prelude. IDL 2.0 models will fail if they target these shapes.
+3. Update the Smithy IDL 2.0 model loader implementation to be able to load
+   Smithy 1.0 models alongside Smithy 2.0 models.
+   1. Warn when a Smithy 1.0 model file is loaded.
+   2. Inject the `@default` trait on structure members when needed.
+   3. Remove the `@box` trait from the model.
+
+
+## `@default` trait
+
+The `@default` trait can be applied to structure members to indicate that the
+targeted shape has a default, zero value. The `@default` trait *does not* allow
+for a custom default value.
+
+The following example defines a structure with a `@default` "title" member that
+has a default zero value:
+
+```
+structure Message {
+    @default
+    title: String // defaults to ""
+}
+```
+
+The `@default` trait is defined in Smithy as:
+
+```
+@trait(selector: """
+       structure > member :not(> union)""",
+       conflicts: [required, box])
+structure default {}
+```
+
+The `@default` trait conflicts with the `@box` trait and `@required` trait. The
+following model is invalid:
+
+```
+structure Message {
+    @default    // ERROR: this trait conflicts with @required.
+    @required
+    title: String
+}
+```
+
+
+### **Default zero values**
+
+The following list describes the default zero value of each kind of shape.
+Programming languages and code generators that cannot initialize structure
+members with the following default values SHOULD continue to represent those
+members as nullable as this is semantically equivalent to the default zero
+value.
+
+- Boolean: boolean false (`false`)
+- Numbers: numeric zero (`0`)
+- String: an empty string (`""`). Strings with the enum trait also have a
+  default value of "".
+- Blob: an empty blob. This includes blob shapes marked as `@streaming`.
+- Timestamp: zero seconds since the epoch (for example, `0`, or
+  `1970-01-01T00:00:00Z`).
+- Document: a null document value.
+- List: an empty list (`[]`).
+- Set: an empty set (`[]`).
+- Map: an empty map (`{}`).
+- Structure: an empty structure. Structures only have a default zero value if
+  none of the members of the structure are marked with the `@required` trait.
```

Comment:
This might get answered later, but doesn't this leave open a door for backwards-incompatible changes when child structures remove the required trait?

Reply:
How so?

I don't think so because you have to opt-in to `@default`, and you can only opt-in to `@default` on a member that targets a structure if the structure has all optional members.